File: TUnit.TestProject/ClassConstructorWithEnumerableTest.cs
using Microsoft.Extensions.DependencyInjection;

namespace TUnit.TestProject;

[ClassConstructor<DependencyInjectionClassConstructor>]
[NotInParallel]
public sealed class ClassConstructorWithEnumerableTest(IServiceProvider services) : IDisposable
{
    private bool _isDisposed;

    [Before(Test)]
    public void Setup()
    {
        if (_isDisposed) 
        {
            throw new ObjectDisposedException(nameof(ClassConstructorWithEnumerableTest));
        }
    }

    [Test]
    [MethodDataSource(nameof(GetValues))]
    public void DoSomething(int value)
    {
        ActivatorUtilities.GetServiceOrCreateInstance<DummyReferenceTypeClass>(services);
    }

    public static IEnumerable<int> GetValues() => [1, 2, 3, 4];

    public void Dispose()
    {
        _isDisposed = true;
    }
}

File: TUnit.TestProject.FSharp/ClassConstructorWithEnumerableTest.fs
namespace TUnit.TestProject.FSharp

open System
open Microsoft.Extensions.DependencyInjection
open TUnit
open TUnit.TestProject
open TUnit.Core

[<ClassConstructor(typeof<DependencyInjectionClassConstructor>)>]
[<NotInParallel>]
type ClassConstructorWithEnumerableTest(services: IServiceProvider) =
    let mutable isDisposed = false

    [<Before(HookType.Test)>]
    member _.Setup() =
        if isDisposed then
            raise (ObjectDisposedException(nameof(ClassConstructorWithEnumerableTest)))

    [<Test>]
    [<MethodDataSource("GetValues")>]
    member _.DoSomething(value: int) =
        ActivatorUtilities.GetServiceOrCreateInstance<DummyReferenceTypeClass>(services) |> ignore

    static member GetValues() : seq<int> = seq { yield 1; yield 2; yield 3; yield 4 }

    interface IDisposable with
        member _.Dispose() =
            isDisposed <- true